frontend/apps/reader/modules/readeruserhyph.lua

```diff
@@ -0,0 +1,248 @@
+    end
+end
+
+-- Reload on change of the hyphenation language
+function ReaderUserHyph:onTypographyLanguageChanged()
+    self:loadUserDictionary()
+end
+
+-- Reload on "ChangedUserDictionary" event,
+-- doesn't load dictionary if filesize and filename haven't changed
+-- if reload==true reload
+function ReaderUserHyph:loadUserDictionary(reload)
+    self:loadDictionary(self:isAvailable() and self:getDictionaryPath() or "", reload and true or false)
+    self.ui:handleEvent(Event:new("UpdatePos"))
+end
+
+-- Functions to use with the UI
+
+function ReaderUserHyph:isAvailable()
+    return G_reader_settings:isTrue("hyph_user_dict") and self:_enabled()
+end
+
+function ReaderUserHyph:_enabled()
+    return self.ui.typography.hyphenation
+end
+
+-- add Menu entry
+function ReaderUserHyph:getMenuEntry()
+    return {
+        text = _("Additional user dictionary"),
+        help_text = _([[The hyphenation rules for individual words can be changed with the user dictionary.
+Words not in that dictionary will be hyphenated by the selected method.
+
+Hyphenation can be changed by long pressing more than three seconds on a word and selecting 'Hyphenate' in the popup menu.
+
+If a word is removed from the user dictionary, the selected hyphenation method is applied again.]]),
+        callback = function()
+            local hyph_user_dict =  not G_reader_settings:isTrue("hyph_user_dict")
+            G_reader_settings:saveSetting("hyph_user_dict", hyph_user_dict)
+            self:loadUserDictionary() -- not needed to force a reload here
+        end,
+        checked_func = function()
+            return self:isAvailable()
+        end,
+        enabled_func = function()
+            return self:_enabled()
+        end,
+        separator = true,
+    }
+end
+
+-- Helper functions for dictionary entries-------------------------------------------
+
+-- checks if suggestion is well formated
+function ReaderUserHyph:checkHyphenation(suggestion, word)
+    if suggestion:find("%-%-") then
+        return false -- two or more consecutive '-'
+    end
+
+    suggestion = suggestion:gsub("-","")
+    if self.ui.document:getLowercasedWord(suggestion) == self.ui.document:getLowercasedWord(word) then
+        return true -- characters match (case insensitive)
+    end
+    return false
+end
+
+function ReaderUserHyph:updateDictionary(word, hyphenation)
+    local dict_file = self:getDictionaryPath()
+    local new_dict_file = dict_file .. ".new"
+    local dict
+    local new_dict
+
+    -- if no file, create an empty one
+    if lfs.attributes(dict_file, "mode") ~= "file" then
+        dict = io.open(dict_file, "w")
+        if dict then
+            dict:close()
+        else
+            return
+        end
+    end
+
+    -- open files
+    dict = io.open(dict_file, "r")
+    if not dict then
+        logger.err("UserHyph: could not open " .. dict_file)
+        return
+    end
```

Comment:
This looks like it can be make one step.

Reply:
I don't see how. The first block creates an empty dictionary, if none exists.
The second block opens the dict (it is certain that it exists then).

I need an open dict file for the rest of the function.

Reply:
Something like:
```lua
function ReaderUserHyph:updateDictionary(word, hyphenation)
    local dict_file = self:getDictionaryPath()
    local new_dict_file = dict_file .. ".new"

    local new_dict = io.open(new_dict_file, "w")
    if not new_dict then
        logger.err("UserHyph: could not open " .. new_dict_file)
        return
    end

    local word_lower = self.ui.document:getLowercasedWord(word)
    local line
    local dict = io.open(dict_file, "r")
    if dict then
        -- look for word in current dict, copying lines into new_dict
        ...
    end

    -- write new entry
    if hyphenation and hyphenation~="" then
        new_dict:write(string.format("%s;%s\n", word, hyphenation))
    end
    -- write old entry if there was one
    if line then
        new_dict:write(line .. "\n")
    end

    if dict then -- copy rest of file
        repeat
            line = dict:read()
            if line then
                new_dict:write(line .. "\n")
            end
        until (not line)
        dict:close()
        os.remove(dict_file)
    end
    new_dict:close()
    os.rename(new_dict_file, dict_file)
```